File: src/main/java/com/example/jpabulkinsert/BookRegisterService.java
package com.example.jpabulkinsert;

import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class BookRegisterService {

    private final AutoIncrementedIdBookRepository autoIncrementedIdBookRepository;
    private final SequenceIdBookRepository sequenceIdBookRepository;
    private final CustomIdBookRepository customIdBookRepository;

    @Transactional
    public List<AutoIncrementedIdBookEntity> registerAutoIncrementedIdBooks(List<BookSaveVo> voList) {

        List<AutoIncrementedIdBookEntity> newBooks = voList.stream()
                .map(AutoIncrementedIdBookEntity::from)
                .collect(Collectors.toList());

        return autoIncrementedIdBookRepository.saveAll(newBooks);
    }

    @Transactional
    public List<SequenceIdBookEntity> registerSequenceIdBooks(List<BookSaveVo> voList) {

        List<SequenceIdBookEntity> newBooks = voList.stream()
                .map(SequenceIdBookEntity::from)
                .collect(Collectors.toList());

        return sequenceIdBookRepository.saveAll(newBooks);
    }

    @Transactional
    public List<CustomIdBookEntity> registerCustomIdBooks(List<BookSaveVo> voList) {

        List<CustomIdBookEntity> newBooks = voList.stream()
                .map(CustomIdBookEntity::from)
                .collect(Collectors.toList());

        return customIdBookRepository.saveAll(newBooks);
    }

}

File: src/main/kotlin/com/example/bookstore/domain/book/service/BookRegisterService.kt
package com.example.bookstore.domain.book.service

import com.example.bookstore.domain.book.dto.BookSaveDto
import com.example.bookstore.domain.book.entity.Book
import com.example.bookstore.domain.book.entity.BookJ
import com.example.bookstore.domain.book.repository.BookRepositoryJ
import lombok.RequiredArgsConstructor
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.math.BigDecimal
import java.util.stream.Collectors

@Service
@RequiredArgsConstructor
@Transactional
class BookRegisterService {
    private val bookRepositoryJ: BookRepositoryJ? = null

    fun register(dtos: List<BookSaveDto?>): List<BookJ> {

        val book = Book(bookSeq = 3L, title = "s", author = "", price = BigDecimal.ZERO)

        val newBookJS = dtos.stream()
            .map { dto: BookSaveDto? -> BookJ.from(dto) }
            .collect(Collectors.toList())


        return bookRepositoryJ!!.saveAll(newBookJS)
    }
}
